### PR TITLE
MAINT: Corrected NumPy API usage in scipy.spatial

### PR DESCRIPTION
**Pull request by czgdp1807**
#### Reference issue

https://github.com/rgommers/scipy/pull/28

### CODE REVIEW

**Comment by czgdp1807** [comment]
ping @rgommers

File: scipy/spatial/src/distance_wrap.c [hunk 1]
```diff
@@ -707,14 +707,14 @@ static PyObject *to_squareform_from_vector_wrap(PyObject *self, PyObject *args)
     return 0;
   }
   NPY_BEGIN_ALLOW_THREADS;
-  n = M_->dimensions[0];
-  elsize = M_->descr->elsize;
+  n = PyArray_DIMS(M_)[0];
+  elsize = PyArray_DESCR(M_)->elsize;
```
Comment (rgommers) [inline scipy/spatial/src/distance_wrap.c:711]: Not a 100% sure without checking that the `->elsize` isn't still deprecated API.

To verify, you can modify `spatial/setup.py` like so (there are examples in other `setup.py` files, like for `optimize`):
```python
from scipy._build_utils import numpy_nodepr_api

config.add_extension(...,
     **numpy_nodepr_api)
```
This is generally useful to do.

Comment (rgommers) [reply scipy/spatial/src/distance_wrap.c:711]: (2 more `->elsize` further down)

Comment (rgommers) [reply scipy/spatial/src/distance_wrap.c:711]: Checking the docs says it's not deprecated API, see `elsize` in https://numpy.org/devdocs/reference/c-api/array.html#c.PyArray_DescrNewFromType. So guess we're good here.

**Review by rgommers — APPROVED** [review]
Thanks @czgdp1807! Looks great, I had just one small doubt. Adding a build flag in `setup.py` should clear that up - may be good to do in future PRs. Since this is ready to go and CI is green, I'll merge it.

**Pull request merged**